MRG: Xdawn

alexandrebarachant opened the pull request:
This is the PR related to #2245.

The code for denoising is inspired from the ICA code.  

Using this [gist](https://gist.github.com/alexandrebarachant/b4d94245db162f6663f7) , you should be able to run the code.

I will add tests after validation of the API

todo list : 
- [x] Covariance estimation and regularization through MNE
- [x] Transform method
- [x] Write tests
- [x] Write examples (denoising and decoding)

agramfort commented:
regarding API. I would pass signal_cov to the init and maybe call apply
transform so it can be passed into a decoding pipeline. If the fit does not
do any dim red then apply is fine.

agramfort commented:
mark the branch as MRG when ready for review.

alexandrebarachant commented:
>  I would pass signal_cov to the init and maybe call apply
transform so it can be passed into a decoding pipeline.

does the decoding pipeline takes epochs as inputs ?
As discussed in #2245, this implementation is closer to an ICA than to the decoding version of Xdawn. the original algorithm is not really compatible with the sklearn API.

> If the fit does not do any dim red then apply is fine.

No dim reduction, this is a Xdawn for denoising.

agramfort commented:
> does the decoding pipeline takes epochs as inputs ?
> 
Yes see csp example
> As discussed in #2245, this implementation is closer to an ICA than to the decoding version of Xdawn. the original algorithm is not really compatible with the sklearn API.
> 
Ok fine with apply then
> If the fit does not do any dim red then apply is fine.
> 
> No dim reduction, this is a Xdawn for denoising.
> 
Ok. Thanks 
> —
> Reply to this email directly or view it on GitHub.
>

alexandrebarachant commented:
> Yes see csp example

CSP seems to takes epochs_data as input, not an epoch instance.
Xdawn needs events position to deal with overlapped responses.

 > Ok fine with apply then

i can add a transform method that does the reduction

agramfort commented:
arfff I think it should accept Epochs...

I'll discuss with @trachelr today as he's currently refactoring the
decoding module.

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,412 @@
+    evo = np.dot(np.dot(linalg.pinv(np.dot(to_tot, to_tot.T)), to_tot), data.T)
+
+    # parse evoked response
+    evoked_data = {}
```
evoked_data = dict()

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,412 @@
+def _least_square_evoked(data, events, event_id, tmin, tmax, sfreq, decim):
+    """Least square estimation of evoked response from data.
+
+    return evoked data and toeplitz matrices.
```
can you document the params especially the shapes of arrays?

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,412 @@
+
+
+def _check_overlapp(epochs):
+    """check if events are overllaped."""
```
overlapped

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,412 @@
+    epochs : Epoch object
+        An instance of Epoch.
+    return_toeplitz : bool
+        if true, compute the toeplitz matrix
```
If true, compute the toeplitz matrix.

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,412 @@
+    evo, to = _least_square_evoked(data, evs, epochs.event_id,
+                                   tmin=epochs.tmin, tmax=epochs.tmax,
+                                   sfreq=epochs.info['sfreq'], decim=1)
+    evokeds = {}
```
dict()

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,412 @@
+
+
+class Xdawn():
+
```
empty line not needed.

alexandrebarachant replied:
According to D203, 1 blank line required before class docstring. 
I can remove it but flake8 will not be happy

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,412 @@
+    """Implementation of the Xdawn Algorithm.
+
+    Xdawn is a spatial filtering method designed to improve the signal
+    to signal + Noise ratio (SSNR) of the ERP responses. Xdawn was originaly
```
Noise -> noise

dengemann replied:
should btw subclass from object

alexandrebarachant replied:
Not sure i understand :( Can you be more specific ?

dengemann replied:
```class Xdawn(object):```

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,412 @@
+        self.filters_ = {}
+        self.patterns_ = {}
+        self.evokeds_ = {}
+        self.projs_ = {}
```
replace {} by dict()

it's for readability

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,412 @@
+        self.evokeds_ = {}
+        self.projs_ = {}
+
+    def fit(self, epochs, signal_cov=None, correct_overlap='auto'):
```
move signal_cov and correct_overlap to init

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,412 @@
+        self : Xdawn instance
+            The Xdawn instance.
+        """
+        if correct_overlap is 'auto':
```
never ever use "is" to compare strings. Use ==

it's a big source of errors.

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,412 @@
+            evo, to = least_square_evoked(epochs, return_toeplitz=True)
+        else:
+            evo = {eid: epochs[eid].average() for eid in epochs.event_id}
+            to = {eid: 1.0 / len(epochs[eid]) for eid in epochs.event_id}
```
dict comprehension will not work on python 2.6

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,412 @@
+            The data to be processed.
+        event_id : dict or None
+            The kind of event to apply. if none, a dict of inst will be return
+            one for each type of event xdawn has been fitted.
```
does this apply just to Epochs? If so I would remove this option and just pass Epochs for the event of interest

agramfort commented:
that's it for now. Looks really cool

thanks @alexandrebarachant  !

trachelr commented:
Cool thanks @alexandrebarachant for this nice contrib!
I get the code and see how we can include it in the decoding API

alexandrebarachant commented:
@trachelr I will push the transform method later today.
If the decoding API will accept epochs instance, it would be straightforward.

agramfort commented:
let us know when it ready for review. Mark the PR as [MRG] when it is.

thanks

trachelr commented:
@alexandrebarachant, I didn't have much time to work on it today.
We'll discuss about the decoding API with @agramfort tomorrow, review and test your code.
Thanks again!

alexandrebarachant commented:
@agramfort . Noted.

A last question. Do you have a nice way (with minimal parameters) to regularize covariance from data (not epochs or raw instance). I would like to do the exact same thing as in the CSP code (L87 to L136) but i don't want to duplicate code. 
The `_compute_covariance_auto` looks good but there is an over complicated list of params to add.

dengemann commented:
>  Do you have a nice way (with minimal parameters) to regularize covariance from data (not epochs or raw instance)

what is the use case? if you have many samples you don't need the regularization.

dengemann commented:
@alexandrebarachant alternatively checkout the grid-searched shrunk cov estimator in cov.py with different reg for sensor types.

alexandrebarachant commented:
@dengemann . Exact same use case as in the CSP.  i would like the possibility to do ledoit_wolf or oas regularization. No need for more fancy method. The spatial filters estimation is based on a generalized eigenvalue decomposition, that could become a bit unstable if the matrices are not well conditioned.

dengemann commented:
> The spatial filters estimation is based on a generalized eigenvalue decomposition, that could become a bit unstable if the matrices are not well conditioned.

they should be roughly equivalent, the shrunk version we have usually outperforms the other oracle estimators as it regularizes channel types individually. But I think here it is not critical. Generally having reg at all makes a big difference.

agramfort commented:
just extract the reg code from csp.py. put it in a private function and
reuse it in Xdawn

alexandrebarachant commented:
Tests and examples added.
Test are more about checking the code is running than verifying it does what is supposed to do.

Ready for review

agramfort commented on examples/decoding/plot_decoding_xdawn_meg.py:
```diff
@@ -0,0 +1,90 @@
+from sklearn.metrics import classification_report, confusion_matrix
+from sklearn.preprocessing import MinMaxScaler
+import numpy as np
+import matplotlib.pyplot as plt
```
import order is not correct. Start from most general to local ie from np, plt to mne

agramfort commented on examples/decoding/plot_decoding_xdawn_meg.py:
```diff
@@ -0,0 +1,90 @@
+
+ERF decoding with Xdawn. For each event type, a set of spatial Xdawn filters
+are trained and apply on the signal. Channels are concatenated and rescaled to
+create features vectors that will be fed into a Logistic Regression.
```
can you add a ref here?

agramfort commented on examples/preprocessing/plot_xdawn_denoising.py:
```diff
@@ -0,0 +1,65 @@
+space and then projected back in the sensor space using only the first two
+xdawn components. The process is similar to an ICA, but is
+supervised in order to maximize the signal to signal + noise ratio of the
+evoked response.
```
add ref

agramfort commented:
I'll continue tomorrow. Can you see what travis is not happy?

dengemann commented on examples/decoding/plot_decoding_xdawn_meg.py:
```diff
@@ -0,0 +1,101 @@
+                add_eeg_ref=False, verbose=False)
+
+# Create classification pipeline
+clf = make_pipeline(Xdawn(3),
```
for readability please use a keyword argument, e.g. `n_components=3`

dengemann commented on examples/decoding/plot_decoding_xdawn_meg.py:
```diff
@@ -0,0 +1,101 @@
+labels = epochs.events[:, -1]
+
+# Cross validator
+cv = StratifiedKFold(labels, 10, shuffle=True, random_state=42)
```
keyword arguments would be nice ;)

dengemann commented on examples/preprocessing/plot_xdawn_denoising.py:
```diff
@@ -0,0 +1,76 @@
+
+# Plot image epoch after xdawn
+plot_image_epochs(epochs_denoised['vis_r'],
+                  picks=[230], vmin=-500, vmax=500)
```
fits in one line?

dengemann commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+            self.signal_cov_ = _regularized_covariance(sig_data, self.reg)
+
+        # estimates evoked covariance
+        self.evokeds_cov_ = {}
```
dict()

dengemann commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+
+            # Create and add projectors
+            ch_names = self.evokeds_[eid].ch_names
+            projs = []
```
list()

dengemann commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+                             'type or numpy array')
+
+        # create full matrix of spatial filter
+        V = []
```
list()

dengemann replied:
and if you can give it a name -- we try to avoid single character names outside of list comprehensions

dengemann commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+
+        picks = pick_types(raw.info, meg=False, include=self.ch_names,
+                           exclude='bads')
+        raws = {}
```
dict

dengemann commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+                               'xdawn.ch_names' % (len(self.ch_names),
+                                                   len(picks)))
+
+        epochs_dict = {}
```
dict()

dengemann commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+                                                        len(picks)))
+
+        data = evoked.data[picks]
+        evokeds = {}
```
dict()

dengemann commented:
@alexandrebarachant 
you seem to have some numerical errors in your tests according to travis.
Maybe not enough data?

```
======================================================================
ERROR: Test Xdawn fit.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/mne-tools/mne-python/mne/preprocessing/tests/test_xdawn.py", line 66, in test_xdawn_fit
    xd.fit(epochs)
  File "/home/travis/build/mne-tools/mne-python/mne/preprocessing/xdawn.py", line 354, in fit
    self.signal_cov_)
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/scipy/linalg/decomp.py", line 420, in eigh
    " computed." % (info-b1.shape[0]))
LinAlgError: the leading minor of order 60 of 'b' is not positive definite. The factorization of 'b' could not be completed and no eigenvalues or eigenvectors were computed.
```

alexandrebarachant commented:
@dengemann That's a weird bug. it pass on my laptop and on appveyor, but not on travis. Probably related to the scipy version.
I guess the test signal is composed of noise only, and it cause trouble to the least square estimation to get well conditioned matrices. I will try to reduce the number of channels. If it doesn't works i will switch the regularization parameters.

dengemann commented:
>  I will try to reduce the number of channels. If it doesn't works i will switch the regularization parameters.

if you want to save some time maybe look at our travis.yml setup file and install in a new environment the exact version which makes it crash

agramfort commented on examples/decoding/plot_decoding_xdawn_meg.py:
```diff
@@ -0,0 +1,101 @@
+
+# Normalized confusion matrix
+cm = confusion_matrix(labels, preds)
+cm_normalized = cm.astype('float') / cm.sum(axis=1)[:, np.newaxis]
```
'float' -> float

is enough

agramfort commented on mne/preprocessing/tests/test_xdawn.py:
```diff
@@ -0,0 +1,137 @@
+    # ========== with cov regularization ====================
+    # ledoit-wolf
+    xd = Xdawn(n_components=2, correct_overlap=False,
+               signal_cov=np.eye(len(picks)), reg='lws')
```
lws? why not lw ?

alexandrebarachant replied:
This come from the CSP code. I kept the same naming convention.

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+
+    Parameters
+    ----------
+    data : array, shape=(n_channels, n_times)
```
replace = by space

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+        If some events don't match the events of interest as specified
+        by event_id, they will be ignored.
+    event_id : dict
+        The id of the event to consider
```
event -> events

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+
+    Parameters
+    ----------
+    data : array, shape=(n_channels x n_times)
```
=

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+                                 "of type str (got %s)." % type(reg))
+        else:
+            raise ValueError("regularization parameter should be "
+                             "of type str (got %s)." % type(reg))
```
factorize error msg

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+
+
+def least_square_evoked(epochs, return_toeplitz=False):
+    """Least square estimation of evoked response from a epoch instance.
```
epoch -> Epochs

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+
+    Parameters
+    ----------
+    epochs : Epoch object
```
Epochs instance

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+    Parameters
+    ----------
+    epochs : Epoch object
+        An instance of Epoch.
```
Epoch -> Epochs

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+        ----------
+        epochs : Epoch object
+            An instance of Epoch on which Xdawn filters will be trained.
+        y : None
```
None | ???

agramfort replied:
oups my bad...

agramfort replied:
well you might still want to specify y...

alexandrebarachant replied:
y is not used in the code, it is to keep compatibility with pipelining with sklearn.
the class are derived from epochs.event_id

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+
+        Returns
+        -------
+        X : array, shape(n_trials x n_components * event_types x n_times)
```
space missing after shape

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+
+        Parameters
+        ----------
+        epochs : Epoch or array, shape=(n_trial x n_channels x n_times)
```
extra =

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+
+            epochs_r = epochs.copy()
+            epochs_r._data[:, picks] = np.array(np.split(data_r,
+                                                len(epochs.events), 1))
```
what is this doing?

alexandrebarachant replied:
data are reconstructed from the concatenation of the epochs data. I guess is to fasten the process.
This is code from ica.py

agramfort commented on examples/decoding/plot_decoding_xdawn_meg.py:
```diff
@@ -0,0 +1,101 @@
+tick_marks = np.arange(len(target_names))
+plt.xticks(tick_marks, target_names, rotation=45)
+plt.yticks(tick_marks, target_names)
+plt.tight_layout()
```
please use mne.viz.tight_layout()

agramfort commented on examples/decoding/plot_decoding_xdawn_meg.py:
```diff
@@ -0,0 +1,101 @@
+plt.tight_layout()
+plt.ylabel('True label')
+plt.xlabel('Predicted label')
+plt.show()
```
how does the CSP compete with Xdawn on the very same data? just for our curiosity.

alexandrebarachant replied:
No idea. 
Direct comparison is difficult. CSP is not multiclass.
the MNE CSP `transform` produce variance, while here we want to classify time samples.

CSP is not build for this task, but it might works well on MEG data because the signal is clean enough to produce a change in variance when the ERP arise.

mainakjas replied:
should we use EEG data then? The sample data has EEG I think. It might be better to demonstrate this algorithm. I fear that what we have currently is "too easy". If you just feed the time samples to an SVM, I think it does just as well for the two class accuracies you report ...

mainakjas commented on examples/decoding/plot_decoding_xdawn_meg.py:
```diff
@@ -0,0 +1,101 @@
+raw_fname = data_path + '/MEG/sample/sample_audvis_filt-0-40_raw.fif'
+event_fname = data_path + '/MEG/sample/sample_audvis_filt-0-40_raw-eve.fif'
+tmin, tmax = -0.1, 0.3
+event_id = dict(aud_l=1, aud_r=2, vis_l=3, vis_r=4)
```
how does the 2 class accuracy compare to the 4 class accuracy?

alexandrebarachant replied:
- aud_l v.s. aud_r  = 82%
- viz_l v.s. viz_r  = 96.5%
- aud_l v.s. viz_l  = 98.6%
- aud_l v.s. viz_r  = 95.7%
- aud_r v.s. viz_l  = 99.3%
- aud_r v.s. viz_r  = 97.2%

the 4 class accuracy is 93.4%

mainakjas replied:
look at this example: http://martinos.org/mne/stable/auto_examples/decoding/plot_decoding_sensors.html#example-decoding-plot-decoding-sensors-py. With just a single time point on SVM, you can get similar accuracies. I fear this is just too easy. We should find harder data for demonstrating this. What do you think?

alexandrebarachant replied:
Indeed, i removed the Xdawn step and the classifier (4 classes) give me the same accuracy.
MEG data are simply too good. 

On the EEG data, 4-classes accuracy goes from 73.2% to 85.4% with xdawn.

Saying that, a well tuned classifier is supposed to be as good as any spatial filtering method, because spatial filtering is a linear step, and can be embedded in the classifier.

What you gain with xdawn is the dimentionality reduction

mainakjas replied:
wait, I missed something. What is 73.2% and what is 85.4% ? 

The rest makes sense. So, what do you think about changing the data to EEG (actually I would vote that this is done for all our decoding examples to make it a slightly harder problem with the data we have)?

alexandrebarachant replied:
By removing Xdawn from the decoding pipeline, perf on EEG data give me 73.2%.
If you add Xdawn, it jump to 85%.

agramfort commented:
+1 for merge when Travis is happy 

If you can rebase and update what's new it's great

mainakjas commented on examples/preprocessing/plot_xdawn_denoising.py:
```diff
@@ -0,0 +1,75 @@
+Xdawn filters are trained from epochs, signal is projected in the sources
+space and then projected back in the sensor space using only the first two
+xdawn components. The process is similar to an ICA, but is
+supervised in order to maximize the signal to signal + noise ratio of the
```
nitpick. signal-to-noise ratio

alexandrebarachant replied:
this is correct, the original paper speak about signal to (signal+noise) ratio.

mainakjas replied:
my bad ...

mainakjas commented on examples/preprocessing/plot_xdawn_denoising.py:
```diff
@@ -0,0 +1,75 @@
+
+# Setup for reading the raw data
+raw = io.Raw(raw_fname, preload=True)
+raw.filter(1, 20, method='iir')  # replace baselining with high-pass
```
? why 20 Hz?

alexandrebarachant replied:
For pure ERP classification (at least in EEG), there is no useful signal > 20Hz.
Accuracy generally improves with low pass filtering.
May be different with MEG signal.

mainakjas commented on mne/preprocessing/tests/test_xdawn.py:
```diff
@@ -0,0 +1,137 @@
+# License: BSD (3-clause)
+
+import os.path as op
+
```
maybe we don't need a blank line after each import

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+    return evokeds
+
+
+class Xdawn(TransformerMixin, ContainsMixin):
```
To avoid me some pain, please add `versionadded` directive and also `See also` :)

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+
+        Parameters
+        ----------
+        epochs : Epoch object
```
arggh ... this won't work in a scikit-learn pipeline with cross validation. You need to make it numpy array till there's a fix in scikit-learn

alexandrebarachant replied:
Impossible, xdawn need event position to work.
As discussed in this PR, and in issue #2245 , this code is more oriented denoising than decoding.
The XDAWN algorithm is not really compatible with the sklearn API where each sample are assumed to be unordered an independant.

You can still pipeline this in sklearn pipeline, but you have to do the CV by hand

mainakjas replied:
that's just a numpy array of the same shape as `n_samples`? Can't you just add it to the matrix `X` ? (or maybe I'm just tired and don't make much sense :)

alexandrebarachant replied:
Actually, i also use tmin, tmax, sfreq etc.
i basically reconstruct a continuous signal from the epochs and the info.

agramfort replied:
Epoch -> Epochs

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+
+        Parameters
+        ----------
+        epochs : Epoch or array, shape (n_trial x n_channels x n_times)
```
epochs object won't work as far as I know ...

alexandrebarachant replied:
It also accept array.

mainakjas commented on examples/decoding/plot_decoding_xdawn_meg.py:
```diff
@@ -0,0 +1,102 @@
+
+# Do cross-validation
+preds = np.empty(len(labels))
+for train, test in cv:
```
we don't really need this loop if we use the scikit-learn cross-validation API, do we?

alexandrebarachant replied:
As mentioned before, the sklearn crossvalidation seems to takes only np.array.

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+    ix_events = epochs.events[:, 0] - epochs.events[0, 0]
+
+    data = np.zeros((ne, ns))
+    for i, ix in enumerate(ix_events):
```
single letter variable names :)

mainakjas replied:
Also, it seems to me what you refer to as `ix` is what we call `idx`

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+
+    ns = end_ix - start_ix
+    ns_epochs = epochs._data.shape[2]
+    ne = epochs._data.shape[1]
```
hula! why are you accessing private stuff?

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+
+def _construct_signal_from_epochs(epochs):
+    """Reconstruct pseudo continuous signal from epochs."""
+    start_ix = (np.min(epochs.events[:, 0])
```
I would call this just `start` and similarly the other one just `stop`. But I can live with the current nomenclature ...

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+              + int(epochs.tmax * epochs.info['sfreq']) + 1)
+
+    ns = end_ix - start_ix
+    ns_epochs = epochs._data.shape[2]
```
`n_epochs` is what we typically use. Private attributes ... @agramfort would say "oh boy" :)

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+    end_ix = (np.max(epochs.events[:, 0])
+              + int(epochs.tmax * epochs.info['sfreq']) + 1)
+
+    ns = end_ix - start_ix
```
`n_samples` not `ns` ?

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+    if not isinstance(epochs, _BaseEpochs):
+        raise ValueError('epochs must be an instance of `mne.Epochs`')
+
+    evs = epochs.events.copy()
```
even though your code is pretty clean and I don't have issues with your variable name if they were isolated code. But I feel if it's part of a repo, it should be as consistent with the repo norms as possible. That makes life easy for a new developer ...

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+    data = _construct_signal_from_epochs(epochs)
+    evo, to = _least_square_evoked(data, evs, epochs.event_id,
+                                   tmin=epochs.tmin, tmax=epochs.tmax,
+                                   sfreq=epochs.info['sfreq'], decim=1)
```
why have the param `decim` if you never use it?

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+            data_r = self._pick_sources(data, include, exclude, eid)
+
+            epochs_r = epochs.copy()
+            epochs_r._data[:, picks] = np.array(np.split(data_r,
```
private method again ...

alexandrebarachant replied:
Do you have a good way to affect data to an existing Epoch instance ?
Note : this is actual code from ICA

agramfort replied:
you should use EpochsArray rather than epochs.copy() and private attribute hack.

it should also be fixed in ICA

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,565 @@
+        for eid in event_id:
+
+            data_r = self._pick_sources(data, include, exclude, eid)
+            evo = evoked.copy()
```
no, we really need better variable names

mainakjas commented on mne/preprocessing/tests/test_xdawn.py:
```diff
@@ -0,0 +1,137 @@
+    assert_raises(ValueError, Xdawn, signal_cov=42)
+
+
+def test_xdawn_fit():
```
why is it that you don't need the `requires_sklearn` decorator anywhere?

dengemann replied:
yo sklearn decorator here. see ica tests for example

alexandrebarachant replied:
Only regularisation need sklearn

mainakjas commented on mne/preprocessing/tests/test_xdawn.py:
```diff
@@ -0,0 +1,137 @@
+               signal_cov=np.eye(len(picks)), reg=2)
+    assert_raises(ValueError, xd.fit, epochs)
+
+    # fit with baseline correction and ovverlapp correction should throw an
```
overlap

mainakjas commented:
You need to make travis happy

mainakjas commented:
That's it from me for now! I'll take a look again once you've addressed the comments. Thanks for the contribution @alexandrebarachant :)

alexandrebarachant commented:
Ho boy ! so much comments ...
Thank you guys for that.

Travis is not happy due to sklearn used for regularization of cov mat. do you have a flag to skip the test in such case ?

mainakjas commented:
:) 

See this @alexandrebarachant : https://github.com/mne-tools/mne-python/search?utf8=%E2%9C%93&q=requires_sklearn&type=Code

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,572 @@
+            if epochs.baseline is not None:
+                raise ValueError('Baseline correction must be None if overlap '
+                                 'correction activated')
+            evo, to = least_square_evoked(epochs, return_toeplitz=True)
```
indeed can you replace the evo by evokeds

what does "to"  mean?

alexandrebarachant commented:
I made a serious code review addressing all your comment and chasing bad variable names and usage of private methods.
Travis is super happy, the branch is rebased and What's new updated. 
Ready for merging

dengemann commented on examples/decoding/plot_decoding_xdawn_eeg.py:
```diff
@@ -0,0 +1,102 @@
+raw.filter(1, 20, method='iir')
+events = read_events(event_fname)
+
+raw.info['bads'] = ['MEG 2443']  # set bad channels
```
not needed as you mute MEG

dengemann commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,566 @@
+import numpy as np
+from scipy import linalg
+import copy as cp
+from distutils.version import LooseVersion
```
import order should be: 1) standard library 2) numpy scipy / other libraries 3) our code -- so move copy up.

dengemann commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,566 @@
+    window = nmax - nmin
+    n_samples = data.shape[1]
+    toeplitz_mat = dict()
+    full_toep = []
```
list()

dengemann commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,566 @@
+                skl_cov = ShrunkCovariance(shrinkage=reg,
+                                           store_precision=False,
+                                           assume_centered=True)
+        elif isinstance(reg, str):
```
for python 3 compat use six.string_types --- look for it the rest of our code

alexandrebarachant replied:
this is your code :) (CSP.py)

dengemann replied:
ping @trachelr ;)

dengemann replied:
Ayeways, while you're at it ...

dengemann commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,566 @@
+                             "of type str or int (got %s)." % type(reg))
+
+        # compute regularized covariance using sklearn
+        covar = skl_cov.fit(data.T).covariance_
```
for consistency call it cov

dengemann commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,566 @@
+    ----------
+    n_components : int, default 2
+        The number of components to decompose M/EEG signals.
+    signal_cov : None, Covariance instance or ndarray
```
None | Covariance | ndarray of float, shape(n_channels, n_channels)

-- in general take a look at the other doc strings and adopt

dengemann commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,566 @@
+        Apply correction for overlaped ERP for the estimation of evokeds
+        responses. if 'auto', the overlapp correction is chosen in function
+        of the events in epochs.events.
+    reg : float, str, None
```
we use | pipes for separating

dengemann commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,566 @@
+                raise ValueError('Baseline correction must be None if overlap '
+                                 'correction activated')
+            evokeds, toeplitz = least_square_evoked(epochs,
+                                                    return_toeplitz=True)
```
this goes in one line no?

alexandrebarachant replied:
not if i add a space after the ',' :(

dengemann commented:
thanks @alexandrebarachant looks clean! Besides my comments have a look at existing docs and checkout the type annotations especially for arrays. Also I think you can gain a few lines by doing your line breaks at the latest possible position before reaching 79 characters.

alexandrebarachant commented:
> Besides my comments have a look at existing docs and checkout the type annotations especially for arrays

Well, this is what i've done :) My mistake was to pick a bad one (CSP.py)

Eric89GXL commented:
Yeah we are transitioning to a more complete documentation standard, sorry
for the ambiguity.

alexandrebarachant commented:
@Eric89GXL No Problem. Can you point a part of the MNE code you consider as super clean. I had hard time to identify your coding standard.

Eric89GXL commented:
I think the most recent thing merged was preprocessing/maxwell.py, so
hopefully that's a good one. I'm on mobile now so I can't look too closely,
if you can wait a few hours a can look.

If you want you could open an issue for updating our code to reflect the
ideal style, and/or an issue to actually document our preferences
somewhere. (Although we might already have an issue for the latter...) We
should mark it for the next release to make sure it gets done soon so that
the next new contributor doesn't face a similar challenge.

wmvanvliet commented:
Just wanted to report that these functions work correctly on my EEG data as well. Thanks @alexandrebarachant!

wmvanvliet commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,564 @@
+
+    Returns
+    -------
+    evokeds_data : dict ofnd array
```
dict of ndarray

wmvanvliet commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,564 @@
+        If not None, allow regularization for covariance estimation
+        if float, shrinkage covariance is used (0 <= shrinkage <= 1).
+        if str, optimal shrinkage using Ledoit-Wolf Shrinkage ('lws') or
+        Oracle Approximating Shrinkage ('oas').
```
please document all defaults

wmvanvliet commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,564 @@
+    epochs : Epochs instance
+        An instance of Epochs.
+    return_toeplitz : bool
+        If true, compute the toeplitz matrix.
```
please document all defaults

wmvanvliet commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,564 @@
+            one for each type of event xdawn has been fitted.
+        include : array_like of int.
+            The indices refering to columns in the ummixing matrix. The
+            components to be kept.
```
please document all defaults. What does `apply` do when `include = None`? What does `apply do when `exclude = None`?

wmvanvliet commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,564 @@
+            The data to be processed.
+        event_id : dict or None
+            The kind of event to apply. if none, a dict of inst will be return
+            one for each type of event xdawn has been fitted.
```
why is `event_id` a dict and not a str?

agramfort commented on doc/source/whats_new.rst:
```diff
@@ -10,6 +10,8 @@ Current
 Changelog
 ~~~~~~~~~
+    - Add Xdawn algorithm for denoising and decoding of ERP/ERF by `Alexandre
+    Barachant`_
```
we add stuff as the bottom usually

mainakjas replied:
Actually, you haven't provided your homepage information. So, the link won't render correctly

dgwakeman replied:
also, please link to the relevant function(s) and class(es).

alexandrebarachant replied:
Noob here...
How do i do that ?

alexandrebarachant replied:
Nevermind, i found it at the end of the page :)

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,564 @@
+
+    Returns
+    -------
+    cov : ndarray, shape (n_channels x n_channels)
```
(n_channels x n_channels) -> (n_channels, n_channels)

the shapes are tuples

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,564 @@
+            self.signal_cov_ = signal_cov
+        else:
+            raise ValueError('signal_cov must be None, a covariance instance '
+                             'or a ndarray')
```
if we respect sklearn conventions the attributes with trailing underscores are added during fit. You should just do self.signal_cov = signal_cov and set self.signal_cov_ during the fit.

agramfort commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,564 @@
+            evals, evecs = linalg.eigh(self.evokeds_cov_[eid],
+                                       self.signal_cov_)
+            evecs = evecs[:, np.argsort(evals)[::-1]]  # sort eigenvectors
+            evecs /= np.apply_along_axis(np.linalg.norm, 0, evecs)
```
nitpick. I would write:

evecs /= np.sqrt(np.sum(evecs ** 2, axis=0))

agramfort commented:
that's it for me.

thx heaps @alexandrebarachant  !

mainakjas commented on mne/preprocessing/tests/test_xdawn.py:
```diff
@@ -0,0 +1,140 @@
+    # with bad shrinkage
+    xd = Xdawn(n_components=2, correct_overlap=False,
+               signal_cov=np.eye(len(picks)), reg=2)
+    assert_raises(ValueError, xd.fit, epochs)
```
do you want to add the directive `run_tests_if_main` at the bottom so that we can profile the timing and memory?

alexandrebarachant replied:
Added

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,564 @@
+
+    Parameters
+    ----------
+    data : ndarray, shape (n_channels x n_times)
```
woops, `(n_channels, n_times)`

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,564 @@
+    n_samples = stop - start
+    epochs_data = epochs.get_data()
+    n_epochs, n_channels, n_times = epochs_data.shape
+    ix_events = epochs.events[:, 0] - epochs.events[0, 0]
```
dumb question, but what does `ix_` stand for?

Eric89GXL replied:
Index would be my guess

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,557 @@
+
+        Parameters
+        ----------
+        epochs : Epoch | ndarray, shape (n_trial, n_channels, n_times)
```
n_epochs not n_trials :)

mainakjas commented on mne/preprocessing/xdawn.py:
```diff
@@ -0,0 +1,557 @@
+
+        Returns
+        -------
+        X : ndarray, shape (n_trials, n_components * event_types, n_times)
```
same here

agramfort commented:
thanks @alexandrebarachant !

agramfort commented:
:beer: :)

agramfort commented:
@alexandrebarachant would you mind refactoring _regularized_covariance so it's used also in csp code? it would avoid some code duplication.

mainakjas commented:
awesome. Thanks @alexandrebarachant for the contrib !

alexandrebarachant commented:
Thanks all :)

@agramfort Why not. where should it go ?

agramfort commented:
either in csp or xdawn.

wmvanvliet commented:
thanks @alexandrebarachant !
